# 01 — Data Exploration
### Business Request 1 (continued): "Before we build anything on this data, understand what we were actually given."

You already loaded this file into a database with `src/load_data.py`. This notebook deliberately goes back to the **raw CSV** to practice a skill every analyst needs independently of any pipeline: interrogating an unfamiliar dataset yourself, before trusting it.

This follows the same sequence taught in the reference Python-for-Analytics slides: import libraries → read the data → `head`/`tail`/`shape`/`dtypes` → `describe`/`info` → `value_counts`/`unique`/`nunique`.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
pd.set_option('display.max_columns', 20)
sns.set_style('whitegrid')

## 1. Read the data

We use a relative path so this notebook runs the same way for every student who clones the repo (see the README's reproducibility section) — never a hardcoded absolute path.

In [2]:
df = pd.read_csv('../data/raw/pizza_sales.csv')
df.shape

(48620, 12)

## 2. First look: head, tail, shape, columns

In [3]:
df.head()

,pizza_id,order_id,pizza_name_id,quantity,order_date,order_time,unit_price,total_price,pizza_size,pizza_category,pizza_ingredients,pizza_name
0,1.0,1.0,hawaiian_m,1.0,1/1/2015,11:38:36,13.25,13.25,M,Classic,"Sliced Ham, Pineapple, Mozzarella Cheese",The Hawaiian Pizza
1,2.0,2.0,classic_dlx_m,1.0,1/1/2015,11:57:40,16.00,16.00,M,Classic,"Pepperoni, Mushrooms, Red Onions, Red Peppers,...",The Classic Deluxe Pizza
2,3.0,2.0,five_cheese_l,1.0,1/1/2015,11:57:40,18.50,18.50,L,Veggie,"Mozzarella Cheese, Provolone Cheese, Smoked Go...",The Five Cheese Pizza
3,4.0,2.0,ital_supr_l,1.0,1/1/2015,11:57:40,20.75,20.75,L,Supreme,"Calabrese Salami, Capocollo, Tomatoes, Red Oni...",The Italian Supreme Pizza
4,5.0,2.0,mexicana_m,1.0,1/1/2015,11:57:40,16.00,16.00,M,Veggie,"Tomatoes, Red Peppers, Jalapeno Peppers, Red O...",The Mexicana Pizza


In [4]:
df.tail()

,pizza_id,order_id,pizza_name_id,quantity,order_date,order_time,unit_price,total_price,pizza_size,pizza_category,pizza_ingredients,pizza_name
48615,48616.0,21348.0,ckn_alfredo_m,1.0,31-12-2015,21:23:10,16.75,16.75,M,Chicken,"Chicken, Red Onions, Red Peppers, Mushrooms, A...",The Chicken Alfredo Pizza
48616,48617.0,21348.0,four_cheese_l,1.0,31-12-2015,21:23:10,17.95,17.95,L,Veggie,"Ricotta Cheese, Gorgonzola Piccante Cheese, Mo...",The Four Cheese Pizza
48617,48618.0,21348.0,napolitana_s,1.0,31-12-2015,21:23:10,12.00,12.00,S,Classic,"Tomatoes, Anchovies, Green Olives, Red Onions,...",The Napolitana Pizza
48618,48619.0,21349.0,mexicana_l,1.0,31-12-2015,22:09:54,20.25,20.25,L,Veggie,"Tomatoes, Red Peppers, Jalapeno Peppers, Red O...",The Mexicana Pizza
48619,48620.0,21350.0,bbq_ckn_s,1.0,31-12-2015,23:02:05,12.75,12.75,S,Chicken,"Barbecued Chicken, Red Peppers, Green Peppers,...",The Barbecue Chicken Pizza


In [5]:
print("Rows, columns:", df.shape)
print("\nColumn names:", list(df.columns))

Rows, columns: (48620, 12)

Column names: ['pizza_id', 'order_id', 'pizza_name_id', 'quantity', 'order_date', 'order_time', 'unit_price', 'total_price', 'pizza_size', 'pizza_category', 'pizza_ingredients', 'pizza_name']


## 3. Data types

**Why this matters:** `order_date` and `order_time` are read in as plain text
(`object`/`str`), not real dates — pandas has no way to know they're dates
unless we tell it. `pizza_id`, `order_id`, and `quantity` are read as
`float64` even though every value is a whole number, because pandas' CSV
reader defaults numeric columns to float. Neither of these is wrong, exactly
— but neither is *usable* yet either. That's the job of Notebook 02.

In [6]:
df.dtypes

pizza_id             float64
order_id             float64
pizza_name_id            str
quantity             float64
order_date               str
order_time               str
unit_price           float64
total_price          float64
pizza_size               str
pizza_category           str
pizza_ingredients        str
pizza_name               str
dtype: object

In [7]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   pizza_id           48620 non-null  float64
 1   order_id           48620 non-null  float64
 2   pizza_name_id      48620 non-null  str    
 3   quantity           48620 non-null  float64
 4   order_date         48620 non-null  str    
 5   order_time         48620 non-null  str    
 6   unit_price         48620 non-null  float64
 7   total_price        48620 non-null  float64
 8   pizza_size         48620 non-null  str    
 9   pizza_category     48620 non-null  str    
 10  pizza_ingredients  48620 non-null  str    
 11  pizza_name         48620 non-null  str    
dtypes: float64(5), str(7)
memory usage: 10.4 MB


## 4. Descriptive statistics

In [8]:
df.describe()

,pizza_id,order_id,quantity,unit_price,total_price
count,48620.000000,48620.000000,48620.000000,48620.000000,48620.000000
mean,24310.500000,10701.479761,1.019622,16.494132,16.821474
std,14035.529381,6180.119770,0.143077,3.621789,4.437398
min,1.000000,1.000000,1.000000,9.750000,9.750000
25%,12155.750000,5337.000000,1.000000,12.750000,12.750000
50%,24310.500000,10682.500000,1.000000,16.500000,16.500000
75%,36465.250000,16100.000000,1.000000,20.250000,20.500000
max,48620.000000,21350.000000,4.000000,35.950000,83.000000


`describe()` only summarizes numeric columns by default. For a quick look at the categorical (text) columns, use `describe(include='object')`:

In [9]:
df.describe(include='object')

/tmp/ipykernel_474/87514550.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


,pizza_name_id,order_date,order_time,pizza_size,pizza_category,pizza_ingredients,pizza_name
count,48620,48620,48620,48620,48620,48620,48620
unique,91,358,16382,5,4,32,32
top,big_meat_s,26-11-2015,12:32:00,L,Classic,"Pepperoni, Mushrooms, Red Onions, Red Peppers,...",The Classic Deluxe Pizza
freq,1811,261,26,18526,14579,2416,2416


## 5. Interrogating categorical columns: `value_counts`, `unique`, `nunique`

These three methods are how you get to know a column you've never seen before.

In [10]:
df['pizza_category'].value_counts()

pizza_category
Classic    14579
Supreme    11777
Veggie     11449
Chicken    10815
Name: count, dtype: int64

In [11]:
df['pizza_size'].unique()

<ArrowStringArray>
['M', 'L', 'S', 'XL', 'XXL']
Length: 5, dtype: str

In [12]:
print("Distinct pizza flavors (pizza_name):", df['pizza_name'].nunique())
print("Distinct SKUs (pizza_name_id):", df['pizza_name_id'].nunique())
print("Distinct orders:", df['order_id'].nunique())

Distinct pizza flavors (pizza_name): 32
Distinct SKUs (pizza_name_id): 91
Distinct orders: 21350


## Your Turn

Answer each question with one line of pandas. Don't look at the SQL answers
from Phase 2 until you've tried — the goal is to see that pandas can answer
the exact same questions SQL can, just with different syntax.

1. How many distinct values are in `pizza_ingredients`? What does that number, compared to `nunique()` of `pizza_name`, tell you about the relationship between the two columns?
2. What are the 5 most common `unit_price` values in the dataset (`value_counts()`)?
3. Use `.describe()` on just the `quantity` column. What is the maximum number of a single pizza ordered in one line-item?

In [13]:
# Your code here


## What we learned

- 48,620 rows, 12 columns, no obviously missing data (confirmed properly in Notebook 02).
- `order_date` and `order_time` are text, not real dates/times yet.
- `pizza_id`/`order_id`/`quantity` are floats that should be integers.
- The data has a clear hierarchy: `pizza_name` (32 flavors) → `pizza_name_id` (91 flavor+size SKUs) → individual `order_details` rows (48,620).

**Next:** `02_data_cleaning.ipynb` investigates data quality properly and fixes what actually needs fixing — no more, no less.